release.py

```diff
@@ -0,0 +1,694 @@
+    The ".devN" part is the dev suffix.
+    If there is no dev suffix, it returns an empty string for the dev suffix.
+
+    Raises:
+        ValueError if the version format is invalid.
+    """
+    match = re.fullmatch(r'(\d+\.\d+\.\d+)(\.dev\d+)?', version)
+    if not match:
+        raise ValueError(f'Invalid version format: {version}')
+    base_version = match.group(1)
+    dev_suffix = match.group(2) or ''
+    return base_version, dev_suffix
+
+
+def update_pyproject_versions(path: Path, version: str, deps: dict[str, str]) -> None:
+    """Update versions in pyproject.toml."""
+    content = path.read_text()
+    updated = re.sub(rf'version = "{VERSION_REGEX}"', f'version = "{version}"', content)
+    for pkg, pkg_version in deps.items():
+        updated = re.sub(rf'{pkg}=={VERSION_REGEX}', f'{pkg}=={pkg_version}', updated)
+    if content == updated:
+        logger.error(f'No changes made to {path}. Check the versions.')
+        exit(1)
+    path.write_text(updated)
+    logger.info(f'Updated {path} to version {version}')
+
+
+def update_ops_version(ops_version: str, testing_version: str):
+    """Update the ops version in version.py and pyproject.toml."""
+    # version.py
+    ops_src_file_path = VERSION_FILES['ops/src']
+    content = ops_src_file_path.read_text()
+    updated = re.sub(
+        rf"^version: str = '{VERSION_REGEX}'$",
+        f"version: str = '{ops_version}'",
+        content,
+        flags=re.MULTILINE,
+    )
+    ops_src_file_path.write_text(updated)
+    logger.info(f'Updated {ops_src_file_path} to version {ops_version}')
+
+    # pyproject.toml, update both ops-scenario and ops-tracing versions.
+    update_pyproject_versions(
+        VERSION_FILES['ops/pyproject'],
+        testing_version,
+        deps={'ops-scenario': testing_version, 'ops-tracing': ops_version},
+    )
+
+
+def update_testing_version(ops_version: str, testing_version: str):
+    """Update the testing pyproject version."""
+    update_pyproject_versions(VERSION_FILES['testing'], testing_version, deps={'ops': ops_version})
+
+
+def update_tracing_version(ops_version: str):
+    """Update the tracing pyproject version."""
+    update_pyproject_versions(VERSION_FILES['tracing'], ops_version, deps={'ops': ops_version})
+
+
+def update_uv_lock():
+    """Update the uv.lock file with the new versions."""
+    subprocess.run(['uv', 'lock'], check=True)  # noqa: S607
+
+
+def parse_scenario_version():
+    """Parse the current scenario version from pyproject.toml."""
+    file_path = VERSION_FILES['testing']
+    content = file_path.read_text()
+    match = re.search(rf'version = "{VERSION_REGEX}"', content)
+    if not match:
+        raise ValueError(f'Could not find version string in {file_path}')
+    version_str = match.group(1)
+    base_version, dev_suffix = parse_version(version_str)
+    return base_version, dev_suffix
+
+
+def get_new_scenario_version_for_release() -> str:
```

Comment:
Can we skip/automate this step entirely? Don't we want the scenario version to always be exactly major+5, like 3.1.2 -> 8.1.2? (For now at least, and we can always adjust the code later if needed.)

Reply:
We cannot skip it all, because the scenario version differs from the ops version.

This function parses the scenario version from `pyproject.toml` and removes the `.dev0` suffix, uses it as a suggested scenario version for the release, and if the user doesn't accept, they can override it by providing a manually input version for scenario. The logic isn't major +5.